### PR TITLE
Prevent client reuse after release

### LINKED ISSUE #50: Prevent reuse of released client instances
I noticed that if I continue making use of a client reference after calling `done()`, like so:
```
pool.connect(function (err, client, done) {
  if (err) {
    return console.error('error fetching client from pool', err);
  }
  function execute(input, cb) {
    client.query('SELECT $1::int AS number', [input], function (err, result) {
      console.log('Current number of connections in use: %s', pool.pool._inUseObjects.length);
      console.log('Current number of available connections: %s', pool.pool._availableObjects.length);
      process.nextTick(function() {
        return cb(err, result);
      });
    });
  }

  async.each([1, 2, 3], function (i, cb) {
    execute(i, function(err) {
      if (err) console.error('Oh noes: %s', err);
      done();
    });
  });
});
```

...that it would release the items to the pool, but still make use of them:
```
Current number of connections in use: 1
Current number of available connections: 0
Current number of connections in use: 0
Current number of available connections: 1
Current number of connections in use: 0
Current number of available connections: 1
```

I would still expect a user of this library to be careful about reusing connection instances that they've released to the pool, but if you were to use an intermediate object reference that prevents direct access to those connection instances, you could invalidate those references by setting a flag on `client.release()` that prevents people from calling things like `client.query(...)`.

### PR DESCRIPTION
**Pull request by kjdelisle**
closes #50 

This will add an intermediate client object (SafeClient) that wraps calls to `.query` and `.release`, with the goal of preventing reuse of objects that have been released back to the pool.
Attempting to call `.query` after `.release` will throw an error.

### CODE REVIEW

File: test/index.js [hunk 1]
```diff
@@ -75,18 +75,24 @@ describe('pool', function () {
...
+        }
+        safeClient = client
+        client._client.testString = 'foo'
         setTimeout(function () {
-          client.emit('error', new Error('on purpose'))
+          client._client.emit('error', new Error('on purpose'))
         }, 10)
       })
       pool.on('error', function (err) {
+        // Allow all of the event handles to run before checking
```
Comment (kjdelisle) [inline test/index.js]: Errant comment, cleaning this up on next push

**Comment by kjdelisle** [comment]
@brianc PTAL :)

**Comment by brianc** [comment]
I really appreciate you trying to address this issue!  I think it's something we can fix, however, your approach might need some adjustment:

The `SafeClient` approach of making a new instance will unfortunately break a bunch of backwards compatibility with things - there are more methods & properties on the client than just `query` and `release` - I think a better approach would be to monkey patch over `query` and `release` on the client & do the check there.  That way a pooled client has additional behavior, but still is an actual instance of a `Client` object.

We already monkey patch the `release` method on to the client, infact:

https://github.com/brianc/node-pg-pool/blob/master/index.js#L125

**Comment by kjdelisle** [comment]
I broke everything out into SafeClient to keep it tidy, but I hadn't thought about things relying on the object type to be "Client". 😞 

Let me see what I can hack together; I'll push it to this branch when it's ready.

**Comment by kjdelisle** [comment]
I'm having a heck of a time trying to make this work by modifying the client instances directly. The main advantage of abstracting it out with the SafeClient object was that the lifecycle of those instances was a single run from `connect` to `release`. 

By modifying the client instances themselves, we're now responsible for tracking locks of some sort on each of the instances to ensure that the current caller actually _owns_ the resource in question, and it seems like `generic-pool@2.x` doesn't support any sort of locking mechanism.

I'm constantly seeing failures in the `emits acquire every time a client is acquired` test where the parallel calls to `pool.connect` and `pool.query` are receiving the same client instances, detecting the mismatched locks and erroring out.

@brianc Do you have any idea if generic-pool@3.x helps solve or mitigate this problem? I'm hesitant to force-push my experiments over top of this branch just yet, since it's a bit of a mess...

**Comment by brianc** [comment]
Unfortunately this is never going to merge at this point & a new tact needs to be taken.

**Pull request closed without merging**